node-agent/workload_manager/src/workload_manager/workload_listener/container_listener.rs

```diff
@@ -0,0 +1,432 @@
+use crate::workload_manager::workload_listener::listener::WorkloadListener;
+
+use futures::StreamExt;
+
+use bollard::container::{CPUStats, StatsOptions};
+use bollard::models::{ContainerStateStatusEnum, HealthStatusEnum};
+use bollard::Docker;
+
+use tokio::runtime::Runtime;
+use tonic::Status;
+
+use proto::agent::{Instance, InstanceStatus, Resource, ResourceSummary, Status as WorkloadStatus};
+
+use std::sync::mpsc::Sender;
+use sysinfo::{System, SystemExt};
+
+#[derive(Clone)]
+pub struct ContainerListener {}
+
+impl ContainerListener {
+    /// Fetches all the needed data and return it in an InstanceStatus
+    ///
+    /// # Arguments
+    ///
+    /// * `container_id` - container's id
+    /// * `instance` - instance's struct
+    /// * `docker_connection` - bollard's Docker Struct
+    pub async fn fetch_instance_status(
+        container_id: &str,
+        instance: &Instance,
+        docker_connection: &Docker,
+    ) -> Result<InstanceStatus, Status> {
+        let stats_options = Some(StatsOptions {
+            stream: true,
+            one_shot: false,
+        });
+
+        let lim = match instance.clone().resource {
+            Some(resource) => match resource.limit {
+                Some(limit) => limit,
+                None => return Err(Status::internal("Could not get limit in Instance struct")),
+            },
+            None => return Err(Status::internal("No ressource in Instance struct")),
+        };
+
+        let container_data_result = docker_connection
+            .inspect_container(container_id, None)
+            .await;
+        let container_resources_opt = docker_connection
+            .stats(container_id, stats_options)
+            .next()
+            .await;
+
+        if container_resources_opt.is_none() {
+            return Err(Status::internal(format!(
+                "Cannot poll stats from workload {} (container {}) ",
+                instance.id, container_id
+            )));
+        };
+
+        let container_resources_result = container_resources_opt.unwrap();
+
+        let container_resources = match container_resources_result {
+            Ok(res) => res,
+            Err(e) => return Err(Status::internal(e.to_string())),
+        };
```

Comment:
Use `map_err`here.

Reply:
I am confused, do you want me to use map_err to replace this entire block or to change the type of `e`